libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -277,22 +405,91 @@ class state_history_log {
+      if(_begin_block == _end_block)
+         return;
+
+      size_t copy_from_pos = get_pos(_begin_block) + sizeof(uint64_t); //skip first copied block's magic
+      //go ahead and clear the pruned flag from the log file. If we fail to completely vacuum well maybe recovery can sort out something
+      log.seek(0);
+      fc::raw::pack(log, ship_magic(ship_current_version));
+      size_t copy_to_pos = log.tellp();
+
+      //may happen if _begin_block is still first block on-disk of log. just erase 4 byte trailer
+      if(_begin_block == _index_begin_block) {
+         log.flush();
+         fc::resize_file(log.get_file_path(), fc::file_size(log.get_file_path()) - sizeof(uint32_t));
+         return;
+      }
+
+      ilog("Vacuuming pruned log ${n}", ("n", name));
+
+      const size_t offset_bytes = copy_from_pos - copy_to_pos;
+      const size_t offset_blocks = _begin_block - _index_begin_block;
+      log.seek_end(0);
+      size_t copy_sz = log.tellp() - copy_from_pos - sizeof(uint32_t); //don't copy trailer in to new unpruned log
+      const uint32_t num_blocks_in_log = _end_block - _begin_block;
+
+      std::vector<char> buff;
+      buff.resize(4*1024*1024);
+
+      auto tick = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now());
+      while(copy_sz) {
+         const size_t copy_this_round = std::min(buff.size(), copy_sz);
+         log.seek(copy_from_pos);
+         log.read(buff.data(), copy_this_round);
+         log.punch_hole(copy_from_pos, copy_from_pos+copy_this_round);
+         log.seek(copy_to_pos);
+         log.write(buff.data(), copy_this_round);
+
+         copy_from_pos += copy_this_round;
+         copy_to_pos += copy_this_round;
+         copy_sz -= copy_this_round;
+
+         const auto tock = std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now());
+         if(tick != tock)
+            ilog("Vacuuming pruned log ${n}, ${b} bytes remaining", ("b", copy_sz)("n", name));
+         tick = tock;
+      }
       log.flush();
+      fc::resize_file(log.get_file_path(), log.tellp());
+
       index.flush();
-      ilog("fork or replay: removed ${n} blocks from ${name}.log", ("n", num_removed)("name", name));
+      {
+         boost::interprocess::mapped_region index_mapped(index, boost::interprocess::read_write);
+         uint64_t* index_ptr = (uint64_t*)index_mapped.get_address();
+
+         for(uint32_t new_block_num = 0; new_block_num < num_blocks_in_log; ++new_block_num) {
+            const uint64_t new_pos = index_ptr[new_block_num + offset_blocks] - offset_bytes;
+            index_ptr[new_block_num] = new_pos;
+
+            if(new_block_num + 1 != num_blocks_in_log)
+               log.seek(index_ptr[new_block_num + offset_blocks + 1] - offset_bytes - sizeof(uint64_t));
+            else
+               log.seek_end(-sizeof(uint64_t));
+            log.write((char*)&new_pos, sizeof(new_pos));
```

Comment:
Hmm... maybe we should have done relative offset positioning for the entry suffixes. Oh well, not feasible to change now.